src/components/invitation/services/invitation.service.js

```diff
@@ -139,7 +292,125 @@
                     });
             }
 
-            InvitationListenerService.addListeners();
+            function getListenerData(userId, event) {
+                var listenerData = {
+                    path: 'users/' + userId + '/invitations/' + event
+                };
+
+                switch (event) {
+                    case self.listeners.USER_TEACHERS:
+                        listenerData.childAddedHandler = userTeachersChildAdded;
+                        listenerData.childRemoveHandler = userTeachersChildRemove;
+                        break;
+                    case self.listeners.NEW_INVITATIONS:
+                        listenerData.childAddedHandler = newInvitationsChildAdded;
+                        listenerData.childRemoveHandler = newInvitationsChildRemove;
+                        break;
+                    case self.listeners.PENDING_CONFIRMATIONS:
+                        listenerData.childAddedHandler = pendingConfirmationsChildAdded;
+                        listenerData.childRemoveHandler = pendingConfirmationsChildRemove;
+                        break;
+                }
+
+                return listenerData;
+            }
+
+            function userTeachersChildAdded(teacher) {
+                if (angular.isDefined(teacher)) {
+                    UserProfileService.getProfileByUserId(teacher.senderUid).then(function (profile) {
+                        teacher.zinkerzTeacher = profile.zinkerzTeacher;
+                        teacher.zinkerzTeacherSubject = profile.zinkerzTeacherSubject;
+
+                        myTeachers[teacher.senderUid] = teacher;
+                        angular.forEach(registerEvents[StudentContextSrv.getCurrUid()][self.listeners.USER_TEACHERS].cb, function (cb) {
+                            if (angular.isFunction(cb)) {
+                                $timeout(function () {
+                                    cb(myTeachers);
+                                });
+                            }
+                        });
+                    });
+                }
+            }
+
+            function userTeachersChildRemove(teacher) {
+                if (angular.isDefined(teacher)) {
+                    delete myTeachers[teacher.senderUid];
+                    angular.forEach(registerEvents[StudentContextSrv.getCurrUid()][self.listeners.USER_TEACHERS].cb, function (cb) {
+                        if (angular.isFunction(cb)) {
```

Comment:
looks like some of the code in all of the functions from to the apply content is similar, maybe make a base function, at least for the if is angular with the timeout. not a must.